// @xmly/xmrep/es/utils/env.js
export var isWin = typeof window !== 'undefined';
/** 运行环境等 */
export var win = isWin ? window : {};
export var loc = win.location || {};
export var nav = isWin ? navigator : {};
export var getScreenInfo = function () {
    var sc = win.screen;
    return sc ? sc.width + "x" + sc.height : '-';
};
export var getViewportInfo = function () {
    var w = win.innerWidth;
    var b = win.document.body;
    return w ? w + "x" + win.innerHeight : (b ? b.clientWidth + "x" + b.clientHeight : '-');
};
export function isDev() {
    var host = loc ? loc.hostname : '';
    var xenv = win.XMLOG_ENV;
    if (xenv === 'dev') {
        return true;
    }
    if (xenv === 'production' || xenv === 'uat') {
        return false;
    }
    if (/test\.\w+/.test(host)) {
        return true;
    }
    if (/^[0-9\.]+|(?:localhost)$/.test(host)) {
        return true;
    }
    return false;
}
export function isUat() {
    var host = loc ? loc.hostname : '';
    if (win.XMLOG_ENV === 'uat') {
        return true;
    }
    if (/uat\.\w+/.test(host)) {
        return true;
    }
    return false;
}
export var isTest = process.env.NODE_ENV !== 'production';
