// @xmly/xmrep/es/utils/ua.js
import { nav } from './env';
var detect = function (pattern) {
    return function () {
        return pattern.test(ua);
    };
};
export var ua = nav.userAgent || '';
export var uaFunc = {
    isAndroid: detect(/android/i),
    isIOS: detect(/(ipad|iphone|ipod)/i),
    isWindows: detect(/window/i),
    isMac: detect(/mac os x/i),
    isChrome: detect(/webkit\W.*(chrome|chromium)\W/i),
    isFirefox: detect(/mozilla.*\Wfirefox\W/i),
    isGecko: detect(/mozilla(?!.*webkit).*\Wgecko\W/i),
    is360se: detect(/360/i),
    isIE: function () {
        if (nav.appName === 'Microsoft Internet Explorer') {
            return true;
        }
        if (detect(/\bTrident\b/)()) {
            return true;
        }
        return false;
    },
    isOpera: detect(/opera.*\Wpresto\W|OPR/i),
    isSafari: detect(/webkit\W(?!.*chrome).*safari\W/i),
    isMobile: detect(/(iphone|ipod|((?:android)?.*?mobile)|blackberry|nokia)/i),
    isWebKit: detect(/webkit\W/i),
    isIting: detect(/iting/i),
    isItingMain: detect(/iting\(main\)/i),
    isWeixin: detect(/micromessenger/i),
    isSinaWeibo: detect(/weibo/i),
    isQQ: detect(/qq/i),
    isQQWeibo: detect(/tencentmicroblog/i),
};
export var browerInfo = (function () {
    if (!uaFunc.isItingMain()) {
        // 信息仅在 sendByNative 时使用，非主 App 环境可以不必解析
        return {};
    }
    var tem;
    var match = ua.match(/(iting|opera|chrome|safari|firefox|msie|trident(?=\/))\/?\s*(\d+)/i) || [];
    if (/trident/i.test(match[1])) {
        tem = /\brv[ :]+(\d+)/g.exec(ua) || [];
        return 'IE ' + (tem[1] || '');
    }
    if (match[1] === 'Chrome') {
        tem = ua.match(/\b(OPR|Edge)\/(\d+)/);
        if (tem != null)
            return tem.slice(1).join(' ').replace('OPR', 'Opera');
    }
    var _a = match[2] ? [match[1], match[2]] : ['Unknown', '0'], agentName = _a[0], agentVersionMajor = _a[1];
    if ((tem = ua.match(/version\/(\d+)/i)) != null) {
        agentVersionMajor = tem[1];
    }
    return { agentName: agentName, agentVersionMajor: agentVersionMajor, userAgent: ua };
})();
