// @xmly/sniffing-sdk/src/types/index.ts
/*
 * @Date: 2020-10-29 10:40:55
 * @LastEditTime: 2021-02-18 13:51:33
 * @Descripttion: ts接口
 */

/**
 * @description: 初始化参数
 * @param {boolean} autoExpo 是否开启曝光埋点
 * @return {boolean} autoPageTracker 是否开启页面周期埋点
 */
export interface Options {
  autoTrack?: Function
  autoExpo?: boolean
  autoPageTracker?: boolean
  isDebugger?: DEBUGGER
  xmrep?: xmrepOptions
  router?: routerOptions
}
export type routerOptions = {
  type: 'browser' | 'hash'
}
type xmrepOptions = {
  start?: any
  params?: any
}
export type DEBUGGER = true | false

/**
 * @description: 埋点注册表
 */
export interface TrackerMap {
  [id: string]: (obj: any, trackParams?: TrackParams) => void
}

/**
 * @description: 处理后埋点表结构
 */
export interface TrackerEvent {
  id: string
  label: string
  pathname?: string
  cb: (obj: any, trackParams?: TrackParams) => void
}

/**
 * @description: 埋点上报数据
 */
export interface TrackParams {
  // ubtId: string | string[]
  evId: string | string[]
  evParams?: object | null
  evTrigger?: string
  // type?: string //"event" "click" "exposure" "visible"
  // bizType?: string // 业务类型
  // logLevel?: number // 1 - high, 2 - medium(默认), 3 - low
}

export interface RouterObject {
  refererPath?: string
  currentPath?: string
  ext: {
    referer?: Location | HashChangeEvent
    current: Location
    action: string
  }
}
