// @xmly/xmrep/es/functions/log/env.js
import { getScreenInfo, win } from '../../utils/env';
import { ua, uaFunc } from '../../utils/ua';
// === XMLOG环境信息 ===
export default function getEnvInfo() {
    var info = {
        ua: ua,
        os: '',
        device: '',
        browser: '',
        resolution: getScreenInfo(),
        trackType: ('ontouchstart' in win ? 'H5' : 'web'),
        // http://gitlab.ximalaya.com/FEA/apm/master/issues/3
        siteType: uaFunc.isIting() ? 'onPage' : 'offPage',
    };
    if (uaFunc.isMobile()) {
        info.device = 'mobile';
        if (uaFunc.isAndroid()) {
            info.os = 'android';
        }
        if (uaFunc.isIOS()) {
            info.os = 'Ios';
        }
        if (uaFunc.isIting()) {
            info.browser = 'iting';
        }
        if (uaFunc.isWeixin()) {
            info.browser = 'weixin';
        }
        if (uaFunc.isSinaWeibo()) {
            info.browser = 'sinaBlog';
        }
        if (uaFunc.isQQ()) {
            info.browser = 'qq';
        }
        if (uaFunc.isQQWeibo()) {
            info.browser = 'qqBlog';
        }
    }
    else {
        info.device = 'pc';
        if (uaFunc.isWindows()) {
            info.os = 'windows';
        }
        if (uaFunc.isMac()) {
            info.os = 'mac';
        }
        if (uaFunc.isChrome()) {
            info.browser = 'chrome';
        }
        if (uaFunc.isFirefox()) {
            info.browser = 'firefox';
        }
        if (uaFunc.isIE()) {
            info.browser = 'ie';
        }
        if (uaFunc.isOpera()) {
            info.browser = 'opera';
        }
        if (uaFunc.isSafari()) {
            info.browser = 'safari';
        }
        if (uaFunc.is360se()) {
            info.browser = '360';
        }
    }
    return info;
}
