// @xmly/xmrep/es/functions/log/refer.js
import { env } from './share';
export function parseReferrer() {
    var engReg = /(baidu|so|sogou|google|bing|sm)/;
    var keyReg = /(?:^|&)(wd|q|query|keyword|word)=([^&]*)(?:&|$)/;
    var refInfo = {
        refr_source: 'other',
        refr_medium: '',
        refr_term: '',
        refr_from: '',
        referer_url: '',
    };
    var referrer = document.referrer;
    var a = document.createElement('a');
    a.href = referrer;
    if (referrer) {
        refInfo.referer_url = a.href;
        var hostname = a.hostname;
        var search = a.search.slice(1);
        var isFromSE = engReg.test(hostname);
        var hostArr = hostname.match(engReg);
        // 如果来自搜索引擎
        if (isFromSE) {
            refInfo.refr_source = 'search';
            refInfo.refr_medium = hostArr ? hostArr[1] : '';
            // 如果 referrer 有查询参数
            if (search) {
                // 如果匹配到keyReg关键词的key，则设置utm_term
                var keyArray = search.match(keyReg);
                if (keyArray) {
                    refInfo.refr_term = decodeURIComponent(keyArray[2]);
                }
                // 拿到from参数
                if (keyArray = search.match(/(?:^|&)(from)=([^&]*)(?:&|$)/)) {
                    refInfo.refr_from = decodeURIComponent(keyArray[2]);
                }
            }
        }
    }
    var os = env.os;
    if (os === 'android' || os === 'ios') {
        switch (env.browser) {
            case 'iting':
                refInfo.refr_source = 'internal';
                refInfo.refr_medium = 'iting';
                break;
            case 'weixin':
                refInfo.refr_source = 'social';
                refInfo.refr_medium = 'weixin';
                break;
            case 'qq':
                refInfo.refr_source = 'social';
                refInfo.refr_medium = 'qq';
                break;
            case 'sinaBlog':
                refInfo.refr_source = 'social';
                refInfo.refr_medium = 'sinaBlog';
                break;
            case 'qqBlog':
                refInfo.refr_source = 'social';
                refInfo.refr_medium = 'qqBlog';
                break;
        }
    }
    return refInfo;
}
