// @xmly/xmrep/es/functions/log/tracker.js
import { merge } from '../../utils/obj';
import { generateUUID, getTimestamp } from '../../utils/tool';
import { sid } from './share';
import report from './report';
import { collectData } from './collect';
import { ss } from '../../utils/storage';
export var temData = {};
var Tracker = /** @class */ (function () {
    function Tracker(commonProps, val) {
        this.init(commonProps, val);
        report.init();
    }
    Tracker.prototype.init = function (commonProps, val) {
        var _a;
        if (typeof commonProps === 'string') {
            this.commonProps = (_a = {},
                _a[commonProps] = val,
                _a);
        }
        else {
            this.commonProps = commonProps;
        }
        for (var name_1 in this.commonProps) {
            if (name_1 === 'item') {
                temData.page = this.commonProps[name_1];
            }
            if (name_1 === 'itemId') {
                temData.pageId = this.commonProps[name_1];
            }
        }
    };
    Tracker.prototype.pageView = function (extraData) {
        var data = this.mergeCommon();
        data = merge(data, extraData);
        data = merge(getTracert(), data);
        data.serviceId = 'pageview';
        data.appName = data.appName || window.XMLOG_APPNAME || 'event';
        this.report(data);
    };
    Tracker.prototype.event = function (extraData) {
        var data = this.mergeCommon();
        // 事件中 如果没有srcPage，则使用当前页的item ,有则使用自定义值
        data.srcPage = data.srcPage || this.commonProps.item;
        data.srcPageId = data.srcPageId || this.commonProps.itemId;
        // 记录点击位置坐标
        temData.module = extraData.srcModule || '';
        temData.postion = extraData.srcPosition || '';
        data = merge(data, extraData);
        data.appName = data.appName || window.XMLOG_APPNAME || 'event';
        this.report(data);
    };
    /** 整合环境变量 */
    Tracker.prototype.mergeCommon = function () {
        var common = merge({
            uuid: sid,
        }, this.commonProps);
        return merge(common, collectData());
    };
    /** 上报 */
    Tracker.prototype.report = function (data) {
        report.send(this.format(data));
    };
    /** 格式化数据 */
    Tracker.prototype.format = function (data) {
        var props = Object.keys(data).reduce(function (p, key) {
            // 清理空值字段
            if (data[key] != null && data[key] !== '') {
                p[key] = data[key];
            }
            return p;
        }, {});
        var res = {
            _key: generateUUID(),
            props: props,
            ts: getTimestamp(),
        };
        return res;
    };
    return Tracker;
}());
export default Tracker;
function getTracert() {
    var info = {
        srcPage: ss.get('srcPage') || '',
        srcPageId: ss.get('srcPageId') || '',
        srcModule: ss.get('srcModule') || '',
        srcPosition: ss.get('srcPosition') || '',
    };
    return info;
}
