// @xmly/xmrep/es/functions/log/utm.js
import { isUrlPar } from '../../utils/url';
import { trim } from '../../utils/obj';
import { cookie } from '../../utils/storage';
export function collectUtmInfo() {
    var referrer = document.referrer;
    var utmSource = trim(isUrlPar('utm_source'));
    var obj = {
        utm_source: '',
        utm_medium: '',
        utm_campaign: '',
        utm_content: '',
        utm_term: '',
        utm_from: '',
    };
    // 有referrer 且非喜马拉雅
    // var referrer = "http://baidu.com/xmlog/doc/?utm_source=souce&utm_medium=pc&utm_campaign=xx&utm_term=dx&utm_content=121"
    // http://192.168.20.33:8090/xmlog/doc/?utm_source=souce&utm_medium=pc&utm_campaign=xx&utm_term=dx&utm_content=121
    if (referrer) {
        populateFromUrl(obj, referrer);
    }
    // url含有utm_source查询参数
    if (utmSource !== '') {
        populateFromUrl(obj);
    }
    if (obj.utm_source) {
        for (var key in obj) {
            var val = obj[key] || '';
            if (obj.hasOwnProperty(key)) {
                cookie.set(key, encodeURIComponent(val || ''));
            }
        }
    }
    cookie.set('x_xmly_traffic', encodeURIComponent(getUtmInfo()));
}
/** 获取营销数据 */
function getUtmInfo() {
    var arr = [
        'utm_source',
        'utm_medium',
        'utm_campaign',
        'utm_content',
        'utm_term',
        'utm_from',
    ];
    for (var i = 0; i < arr.length; i++) {
        var item = arr[i];
        var val = decodeURIComponent(cookie.get(item));
        if (i === 0 && val == null) {
            return '';
        }
        arr[i] = item + ":" + (val == null ? '' : val);
    }
    if (arr[0]) {
        return arr.join('&');
    }
    return '';
}
/** 默认从当前url去解析 */
function populateFromUrl(model, url) {
    var searchStr;
    if (url) {
        var a = document.createElement('a');
        a.href = url;
        searchStr = a.search;
    }
    for (var p in model) {
        model[p] = decodeURIComponent(isUrlPar(p, searchStr));
    }
    // from 参数格式特殊
    model['utm_from'] = decodeURIComponent(isUrlPar('from', searchStr));
}
