// @xmly/xmrep/es/share/clean.js
import { maxApmValidTime } from '../config';
/**
 * 判断是否http(s)协议
 */
export var isHttp = function (url) {
    return /https?/.test(url);
};
/**
 * 去除url查询参数、去除末尾/
 */
export var cleanUrl = function (url) {
    return url.replace(/([^?]*)(\?|!).+/, '$1').replace(/\/$/, '');
};
/**
 * 去除url中的数字，除非带有v|V|app
 */
export var cleanUrlNumber = function (url) {
    return url.replace(/\/[^/vV(app)]+\d+/g, '');
};
// 去除url末尾数字
export var cleanUrlNumberEnd = function (url) {
    return url.replace(/(\/()\d+)+/, '');
};
/**
 * 删除无用time数据，避免干扰(小于等于0或大于最大值)
 */
export var cleanTime = function (data) {
    for (var key in data) {
        var val = data[key];
        if (typeof val === 'number') {
            data[key] = val | 0;
            if (isNaN(data[key]) || data[key] <= 0 || data[key] >= maxApmValidTime) {
                delete data[key];
            }
        }
    }
};
/** 过滤在白名单内url（不上报） */
export var cleanWhiteList = function (url, whiteList) {
    if (!url || !url.indexOf) {
        return true;
    }
    var isInWhiteList = false;
    for (var i = 0; i < whiteList.length; i++) {
        var it_1 = whiteList[i];
        if (typeof it_1 === 'string') {
            isInWhiteList = !!~url.indexOf(it_1);
        }
        else if (it_1.test) {
            isInWhiteList = it_1.test(url);
        }
        if (isInWhiteList) {
            break;
        }
    }
    return isInWhiteList;
};
