// @xmly/xmrep/es/share/iting.js
import { uaFunc, ua } from '../utils/ua';
import { ss } from '../utils/storage';
import { PRE_PAGE_KEY, PRE_MODULE_KEY } from './pageData';
import { compareVersion } from '../utils/tool';
var iframe;
function loadUrl(url) {
    if (iframe) {
        document.documentElement.removeChild(iframe);
    }
    iframe = document.createElement('iframe');
    iframe.setAttribute('width', '0');
    iframe.style.display = 'none';
    iframe.setAttribute('src', url);
    document.documentElement.appendChild(iframe);
    setTimeout(function () {
        iframe && document.documentElement.removeChild(iframe);
        iframe = null;
    }, 0);
}
function oldVersion(now, com) {
    var nowArr = now.split('.');
    var comArr = com.split('.');
    if (!+nowArr[0] || !+comArr[0])
        return false;
    if (+nowArr[0] < +comArr[0])
        return true;
    if (+nowArr[0] > +comArr[0])
        return false;
    return oldVersion(nowArr.slice(1).join('.'), comArr.slice(1).join('.'));
}
// ios 6.6.9之前版本可能会闪退
var bugVersions = function () {
    var matches = ua.match(/iting.+(\d{1,}\.\d{1,}\.\d{1,}).+ios/);
    if (!uaFunc.isIOS() || !matches) {
        return false;
    }
    return oldVersion(matches[1], '6.6.9');
};
var getVersion = function () {
    return (ua.match(/iting.+(\d{1,}\.\d{1,}\.\d{1,})/) || [, null])[1];
};
/** 永远与native同步ubt中seq、currPage等信息 */
export function notifyNative(currPage, srcModule) {
    if (srcModule === void 0) { srcModule = ''; }
    if (!uaFunc.isIting() || bugVersions()) {
        return;
    }
    // iOS 版本低于 6.7.9 弹屏内的页面同步信息会导致弹屏关闭
    // http://thoughts.ximalaya.com/workspaces/5dc4c8a54cfb7900134fcc8f/docs/5f3df1fef4c00000012a7ac3#5f3e10d2595de2f511cc38f6
    if (uaFunc.isIOS()) {
        var appVersion = getVersion();
        if (appVersion !== null && compareVersion(appVersion, '6.7.9') < 0) {
            return;
        }
    }
    // 本地缓存  防止H5跳H5丢失prePage
    if (currPage) {
        ss.set(PRE_PAGE_KEY, currPage);
    }
    if (srcModule) {
        ss.set(PRE_MODULE_KEY, srcModule);
    }
    var query = encodeURIComponent("currPage=" + currPage + "&srcModule=" + srcModule);
    var schema = "iting://sync.xmly.ubt?_ka=1&data=" + query;
    loadUrl(schema);
}
export function canSendByNative() {
    // 非主 App 环境下
    if (!uaFunc.isItingMain()) {
        return false;
    }
    var appVersion = getVersion();
    // 解析版本异常
    if (appVersion === null) {
        return false;
    }
    // 主 App 支持接口版本
    // http://thoughts.ximalaya.com/workspaces/5e58afa52d54160012c3eab5/docs/5ed61b56f4c0000001d632e3
    if (uaFunc.isIOS() || uaFunc.isAndroid()) {
        return compareVersion(appVersion, '6.7.12') >= 0;
    }
    return false;
}
/**
 * 通过 Blob 计算字符串在 utf-8 编码下的体积
 * @param {string} str
 * @returns
 */
function sizeOf(str) {
    return typeof Blob !== "undefined"
        ? new Blob([str], { type: "text/plain" }).size
        : str.length;
}
export function sendByNative(sessionId, type, subType, common, data) {
    var commonStr = encodeURIComponent(JSON.stringify(common));
    var logStr = encodeURIComponent(JSON.stringify(data));
    var url = "time=" + +new Date() + "&&type=" + type + "&&subType=" + subType + "&&sId=" + (sessionId || 1) + "&&c=" + commonStr + "&&logstr=" + logStr;
    var schema = "iting://sync.xmly.xlog?_ka=1&data=" + encodeURIComponent(url);
    // URL 长度超过 4k 会发生截断
    if (sizeOf(schema) > 4096) {
        return false;
    }
    loadUrl(schema);
    return true;
}
