// @xmly/xmrep/es/config.js
import { __spreadArrays } from "tslib";
import { isDev, isUat } from './utils/env';
import apiWhiteList from './constants/apiWhiteList';
var prodReportOrigin = 'https://mermaid.ximalaya.com';
var testReportOrigin = 'https://mermaid.test.ximalaya.com';
var uatReportOrigin = 'https://mermaid.uat.ximalaya.com';
var reportOrigin = isDev() ? testReportOrigin : isUat() ? uatReportOrigin : prodReportOrigin;
export var binaryReportUrl = reportOrigin + "/collector-web/web-pl/v1";
export var base64ReportUrl = reportOrigin + "/collector-web/web-pl/v2";
// TODO: remove XLOG_URL at 2.0.0
export var finalBRUrl = (typeof window !== 'undefined' && window.XLOG_URL)
    ? window.XLOG_URL
    : binaryReportUrl;
// 错误上报地址
export var reportErrorUrl = '';
// 上报通用type
export var repTypes = {
    APM: {
        name: 'apm_web',
        // 优先级
        priority: 2,
        // 上报开关key
        switchKey: 'apm',
        subTypes: {
            /** 页面性能 */
            PAGE: 'page',
            /** 网络请求时间 */
            NETWORK: 'network',
            /** 资源加载监控 */
            STATICRESOURCE: 'staticResource'
        },
    },
    SHARE: {
        /** 分享跟踪 */
        name: 'share_web',
        // 优先级大于100会持久化缓存
        priority: 102,
        switchKey: 'share',
        subTypes: {
            /** 分享状态 */
            STATE: 'share_stat_record',
            /** 分享回流 */
            FEEDBACK: 'share_feedback',
        },
    },
    UBT: {
        name: 'ubt_web',
        switchKey: 'ubt',
        priority: 101,
        subTypes: {
            /** 页面曝光 */
            PAGE_VIEW: 'pageview',
            /** 页面离开 */
            PAGE_EXIT: 'pageExit',
            /** 点击跳转 */
            CLICK: 'click',
            /** 点击控件 */
            CLICK_BUTTON: 'clickButton',
            /** 控件滑动曝光*/
            SLIP_PAGE: 'slipPage',
            /** 弹框展示 */
            DIALOG_VIEW: 'dialogView',
            /** 弹框点击 */
            DIALOG_CLICK: 'dialogClick',
        },
    },
};
// api监控白名单
export var defaultMonitorWhiteList = __spreadArrays(['xdcs-collector.ximalaya.com', finalBRUrl, binaryReportUrl, base64ReportUrl], apiWhiteList);
// 网络统计刷新间隔
export var networkInterval = 5 * 1000;
// 性能、网络统计最大有效时间
export var maxApmValidTime = 2 * 60 * 1000;
// 上报失败次数限制
export var maxAllowRetryTimes = 3;
// 最多允许合并上传数量
export var maxAllowSendSize = 5;
// 合并请求的时间间隔
export var mergeRequestDuration = 1000;
// 最多允许本地缓存数量
export var maxAllowCacheSize = 20;
// 采样率
export var defaultSampleRate = isDev() ? 1 : 0.5;
// api上报个数
export var maxMonitorQueueLength = 5;
// 上报开关key;
// 通过xlog.write进行业务数据的写关闭，比如 `apm,ubt`，如果没有则为`__null__` （注意前后两个下划线）
// 通过xlog.send进行数据上报的关闭 ，比如 `true`表示发送关闭， `false`表示不关闭
export var reportSwitchKey = 'xlog.send';
export var reportWriteKey = 'xlog.write';
// 获取native信息的key
export var nativeInfoCookieKey = 'xmly_ubt';
// 缓存唯一设备key;
export var sessionIdKey = '_xmLog';
// http://thoughts.ximalaya.com/workspaces/5d231b0cbe825bee8c4d9069/docs/5f2bb6cef4c0000001105761
// export const sessionIdKey = isDev() ? '4&_device' : isUat() ? '6&_device' : '1&_device';
// sdk相关
export var sdkName = 'xmrep';
// 版本
export var sdkVersion = '2.2.12';
